### PR TITLE
feat: Field aggregation

### PR DESCRIPTION
**Pull request by DarkPhoenix2704**
## Change Summary

Closes https://github.com/nocodb/nocodb/issues/7100

## Change type

- [x] feat: (new feature for the user, not a new feature for build script)

### CODE REVIEW

File: packages/nc-gui/composables/useViewAggregate.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+        return {
+          value: aggregations.value[f.field.title] ?? null,
+          field: gridField,
+          column: f.field,
+          index: f.index,
+          width: `${Number(gridField.width.replace('px', ''))}px` || '180px',
+        }
+      })
+    })
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/composables/useViewAggregate.ts:42]: the fields won't change frequently

[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/composables/useViewAggregate.ts:42]: ignore

File: packages/nc-gui/utils/aggregationUtils.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    return `${value}%`
+  }
+
+  if (column.uidt === UITypes.Duration) {
+    return convertMS2Duration(value, parseProp(column.meta)?.duration || 0)
+  }
+
+  console.log(value)
+  return value ?? '∞'
+}
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/utils/aggregationUtils.ts]: ignore

File: packages/nocodb/src/services/data-table.service.ts [hunk 3]
```diff
@@ -65,6 +65,36 @@ export class DataTableService {
...
+    const listArgs: any = { ...param.query };
+
+    try {
+      listArgs.filterArr = JSON.parse(listArgs.filterArrJson);
+    } catch (e) {}
+
+    const data = await baseModel.aggregate(listArgs);
+
+    return data;
+  }
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nocodb/src/services/data-table.service.ts]: ignore

File: packages/nc-gui/composables/useData.ts [hunk 4]
```diff
@@ -446,6 +452,8 @@ export function useData(args: {
       viewId: viewMetaValue.id,
     })
 
+    await reloadAggregate?.trigger()
+
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/composables/useData.ts:456]: ignore

File: packages/nc-gui/composables/useData.ts [hunk 5]
```diff
@@ -522,6 +530,8 @@
       encodeURIComponent(id),
     )
 
+    await reloadAggregate?.trigger()
+
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/composables/useData.ts:534]: ignore

File: packages/nc-gui/composables/useData.ts [hunk 6]
```diff
@@ -76,6 +78,8 @@
         { ...insertObj, ...(ltarState || {}) },
       )
 
+      await reloadAggregate?.trigger()
+
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/composables/useData.ts:82]: ignore

File: packages/nc-gui/composables/useData.ts [hunk 7]
```diff
@@ -214,6 +218,7 @@
         //   query: { ignoreWebhook: !saved }
         // }
       )
+      await reloadAggregate?.trigger({ field: [property] })
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/composables/useData.ts:221]: there is a null check there. Ignore

File: packages/nc-gui/utils/aggregationUtils.ts [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    return `${roundTo(value, 1)}%`
+  }
+
+  if (column.uidt === UITypes.Duration) {
+    return convertMS2Duration(value, parseProp(column.meta)?.duration || 0)
+  }
+
+  return roundTo(value, 1) ?? '∞'
+}
+
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/utils/aggregationUtils.ts:127]: ignore. Let ie be like now

File: packages/nc-gui/composables/useViewAggregate.ts [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+      isPublic,
+      updateAggregate,
+      getAggregations,
+      displayFieldComputed,
+      visibleFieldsComputed,
+    }
+  },
+)
+
+export { useProvideViewAggregate }
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/composables/useViewAggregate.ts:142]: ignore. This composable encapsulated all things related to the aggregations

File: packages/nc-gui/composables/useData.ts [hunk 10]
```diff
@@ -355,6 +360,7 @@ export function useData(args: {
     }
 
     await $api.dbTableRow.bulkUpdate(NOCO, metaValue?.base_id as string, metaValue?.id as string, updateArray)
+    await reloadAggregate?.trigger({ field: props })
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/composables/useData.ts:363]: ignore. There is a null check!!!!!!!!!!!!!!!!!

File: packages/nc-gui/composables/useData.ts [hunk 11]
```diff
@@ -30,6 +30,8 @@
 
   const { isPaginationLoading } = storeToRefs(useViewsStore())
 
+  const reloadAggregate = inject(ReloadAggregateHookInj)
+
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/composables/useData.ts:34]: ignore. There is a null check!!!

File: packages/nc-gui/composables/useData.ts [hunk 12]
```diff
@@ -877,6 +887,7 @@
       const bulkDeletedRowsData = await $api.dbDataTableRow.delete(metaValue?.id as string, rows.length === 1 ? rows[0] : rows, {
         viewId: viewMetaValue?.id as string,
       })
+      await reloadAggregate?.trigger()
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/composables/useData.ts:890]: ignore. There is a null check

File: packages/nocodb/src/db/BaseModelSqlv2.ts [hunk 13]
```diff
@@ -183,10 +184,10 @@
  */
 class BaseModelSqlv2 {
   protected _dbDriver: XKnex;
-  protected model: Model;
   protected viewId: string;
   protected _proto: any;
   protected _columns = {};
+  public model: Model;
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nocodb/src/db/BaseModelSqlv2.ts:190]: ignore

File: packages/nocodb/src/controllers/data-table.controller.ts [hunk 14]
```diff
@@ -117,6 +117,21 @@ export class DataTableController {
...
+    @Req() req: NcRequest,
+    @Param('modelId') modelId: string,
+    @Query('viewId') viewId: string,
+  ) {
+    return await this.dataTableService.dataAggregate(context, {
+      query: req.query,
+      modelId,
+      viewId,
+    });
+  }
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nocodb/src/controllers/data-table.controller.ts:133]: errors are handled inside BaseModelSqlv2

File: packages/nocodb/src/models/GridViewColumn.ts [hunk 15]
```diff
@@ -161,6 +163,7 @@ export default class GridViewColumn implements GridColumnType {
       'group_by',
       'group_by_order',
       'group_by_sort',
+      'aggregation',
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nocodb/src/models/GridViewColumn.ts:166]: ignore

File: packages/nocodb/src/services/public-datas.service.ts [hunk 16]
```diff
@@ -101,6 +101,51 @@ export class PublicDatasService {
...
+    try {
+      listArgs.filterArr = JSON.parse(listArgs.filterArrJson);
+    } catch (e) {}
+
+    try {
+      listArgs.aggregation = JSON.parse(listArgs.aggregation);
+    } catch (e) {}
+
+    return await baseModel.aggregate(listArgs);
+  }
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nocodb/src/services/public-datas.service.ts:147]: ignore them. It was supposed to be like that

File: packages/nocodb/src/services/data-table.service.ts [hunk 17]
```diff
@@ -65,6 +65,44 @@ export class DataTableService {
...
+    } catch (e) {}
+
+    try {
+      listArgs.aggregation = JSON.parse(listArgs.aggregation);
+    } catch (e) {}
+
+    const data = await baseModel.aggregate(listArgs);
+
+    return data;
+  }
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nocodb/src/services/data-table.service.ts:104]: ignore

**Review by dstala — APPROVED** [review]

**Pull request merged**